test/unittest/unittest_nntrainer_models.cpp

```diff
@@ -0,0 +1,359 @@
+                  unsigned int iterations);
+
+private:
+  void readIteration(std::ifstream &f);
+
+  nntrainer::NeuralNetwork nn;
+  WatchedFlatGraph nodes;
+  NodeWatcher loss_node;
+  float expected_loss;
+};
+
+void NodeWatcher::read(std::ifstream &in) {
+  expected_input.read(in);
+
+  /// @note below is harrasing the fact the tensor shares same base memory
+  /// it should better be getVariableRef() or somewhat equivalent
+  for (auto &i : expected_weights) {
+    i.getVariable().read(in);
+  }
+
+  for (auto &i : expected_weights) {
+    i.getGradient().read(in);
+  }
+
+  expected_output.read(in);
+  expected_dx.read(in);
+}
+
+void NodeWatcher::verifyWeight(const std::string &error_msg) {
+  for (unsigned int i = 0; i < expected_weights.size(); ++i) {
+    verify(node->weightAt(i).getVariable(), expected_weights[i].getVariable(),
+           error_msg + " " + node->weightAt(i).getName() + "weight");
+  }
+}
+
+void NodeWatcher::verifyGrad(const std::string &error_msg) {
+  for (unsigned int i = 0; i < expected_weights.size(); ++i) {
+    verify(node->weightAt(i).getGradient(), expected_weights[i].getGradient(),
+           error_msg + " " + node->weightAt(i).getName() + "grad");
+  }
+}
+
+nntrainer::sharedConstTensor
+NodeWatcher::forward(nntrainer::sharedConstTensor in, int iteration) {
+  std::stringstream ss;
+  ss << "forward failed at " << node->getName() << " at iteration "
+     << iteration;
+  std::string err_msg = ss.str();
+
+  verify(*in, expected_input, err_msg + " at input ");
+  verifyWeight(err_msg);
+  nntrainer::sharedConstTensor out = node->forwarding(in);
+  verifyWeight(err_msg);
+  verify(*out, expected_output, err_msg + " at output ");
+  return out;
+}
+
+nntrainer::sharedConstTensor
+NodeWatcher::lossForward(nntrainer::sharedConstTensor pred,
+                         nntrainer::sharedConstTensor answer, int iteration) {
+  std::stringstream ss;
+  ss << "loss failed at " << node->getName() << " at iteration " << iteration;
+  std::string err_msg = ss.str();
+
+  // verify(*pred, expected_input, err_msg + " pred");
+  nntrainer::sharedConstTensor out =
+    std::static_pointer_cast<nntrainer::LossLayer>(node)->forwarding(pred,
+                                                                     answer);
+  // verify(*out, expected_output, err_msg + " out");
+
+  return out;
+}
+
+nntrainer::sharedConstTensor
+NodeWatcher::backward(nntrainer::sharedConstTensor deriv, int iteration,
+                      bool should_verify) {
+  std::stringstream ss;
+  ss << "backward failed at " << node->getName() << " at iteration "
+     << iteration;
+  std::string err_msg = ss.str();
+
+  nntrainer::sharedConstTensor out = node->backwarding(deriv, iteration);
+
+  if (should_verify) {
+    verify(*out, expected_dx, err_msg);
+    verifyGrad(err_msg);
+  }
+
+  auto opt = node->getOptimizer();
+  if (opt) {
+    opt->apply_gradients(node->getWeights(), node->getNumWeights(), iteration);
+  }
+
```

Comment:
missing verifyWeight?

Reply:
Weights are matched here(https://github.com/nnstreamer/nntrainer/pull/653#discussion_r508996995) because the watcher needs to read the next iteration to match the updated weight. IMO, the structure (and format of the golden data) need quite an improvement, but let me do this after making some working version.

Reply:
When are the updated weights of the last iteration verified?

Reply:
Added after loop (line 320)

Reply:
refer to https://github.com/nnstreamer/nntrainer/pull/653/files#r509012565